crates/datasources/src/common/url.rs

```diff
@@ -0,0 +1,129 @@
+            }
+            Err(e) => return Err(DatasourceCommonError::InvalidUrl(e.to_string())),
+            Ok(u) => u,
+        };
+
+        let ds_url = match ds_url.scheme() {
+            "file" => match ds_url.to_file_path() {
+                Ok(f) => Self::File(f),
+                Err(()) => {
+                    return Err(DatasourceCommonError::InvalidUrl(format!(
+                        "url not a valid file: {ds_url}"
+                    )))
+                }
+            },
+            "gs" | "s3" => Self::Url(ds_url),
+            other => {
+                return Err(DatasourceCommonError::InvalidUrl(format!(
+                    "unsupported scheme '{other}'"
+                )))
+            }
+        };
+
+        Ok(ds_url)
+    }
+
+    pub fn scheme(&self) -> DatasourceUrlScheme {
+        match self {
+            Self::File(_) => DatasourceUrlScheme::File,
+            Self::Url(u) => match u.scheme() {
+                "gs" => DatasourceUrlScheme::Gcs,
+                "s3" => DatasourceUrlScheme::S3,
+                _ => unreachable!(),
+            },
+        }
+    }
+
+    pub fn path(&self) -> Cow<str> {
+        match self {
+            Self::File(p) => p.to_string_lossy(),
+            Self::Url(u) => u.path().into(),
+        }
+    }
+
+    pub fn host(&self) -> Option<&str> {
+        match self {
+            Self::File(_) => None,
+            Self::Url(u) => u.host_str(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_url_parse() {
+        let u = DatasourceUrl::new("gs://my_bucket/my_obj").unwrap();
+        assert_eq!(Some("my_bucket"), u.host());
+        assert_eq!("/my_obj", u.path());
+        assert_eq!(DatasourceUrlScheme::Gcs, u.scheme());
+
+        let u = DatasourceUrl::new("gs://my_bucket/my_obj.parquet").unwrap();
+        assert_eq!(Some("my_bucket"), u.host());
+        assert_eq!("/my_obj.parquet", u.path());
+        assert_eq!(DatasourceUrlScheme::Gcs, u.scheme());
+
+        let u = DatasourceUrl::new("./my_bucket/my_obj.parquet").unwrap();
+        assert_eq!(None, u.host());
+        assert_eq!("./my_bucket/my_obj.parquet", u.path());
+        assert_eq!(DatasourceUrlScheme::File, u.scheme());
+
+        let u = DatasourceUrl::new("/Users/mario/my_bucket/my_obj").unwrap();
+        assert_eq!(None, u.host());
+        assert_eq!("/Users/mario/my_bucket/my_obj", u.path());
+        assert_eq!(DatasourceUrlScheme::File, u.scheme());
+
+        let u = DatasourceUrl::new("file:/my_bucket/my_obj.parquet").unwrap();
+        assert_eq!("/my_bucket/my_obj.parquet", u.path());
+        assert_eq!(DatasourceUrlScheme::File, u.scheme());
+
+        // TODO: Maybe we don't want this...
+        let u = DatasourceUrl::new("file:my_bucket/my_obj.parquet").unwrap();
+        assert_eq!("/my_bucket/my_obj.parquet", u.path());
+        assert_eq!(DatasourceUrlScheme::File, u.scheme());
```

Comment:
Whenever creating a URL with `file:` scheme, it assumes the path to be absolute. Maybe we don't want this?

Reply:
The assumption that the path is absolute when using `file:` seems reasonable to me.